Add early stopping utils (#1545)
FluxML/Flux.jl: docs/src/utilities.md

Flux.throttle
Flux.stop
Flux.skip
Flux.patience
Flux.plateau
```

> 

Reviewer: This should be in its own sub-section, separate from the other callback helpers.

Also, the docs reads as a continuation of the docstrings. They should read independently of the docstrings. I would structure the content as "tutorial/docs" then the docstring too.

Flux.stop
Flux.skip
```

## Patience Helpers

> Flux provide utilities for scheduling your workflow according to some monitored condition and a maximum `patience`. For example, you can use `early_stopping` to stop training when the model is converging or deteriorating, or you can use `plateau` to check if the model is stagnating.

Reviewer: ```suggestion
Flux provides utilities for controlling your training procedure according to some monitored condition and a maximum `patience`. For example, you can use `early_stopping` to stop training when the model is converging or deteriorating, or you can use `plateau` to check if the model is stagnating.
```

> 

Reviewer: ```suggestion

For example, below we create a pseudo-loss function that decreases, bottoms out, then increases. The early stopping trigger will break the loop before the loss increases too much.
```julia
# create a pseudo-loss that decreases for 4 calls, then stars increasing
# we call this like loss()
loss = let t = 0
  () -> begin
    (t - 4)^2
    t += 1
  end
end

# create an early stopping trigger
# returns true when the loss increases for two consecutive steps
es = Flux.early￼￼￼_stopping(loss, 2)

# this loop will break early
for iter in 1:100
  @info "Iteration $iter"
  es() && break
end
```

```

The keyword argument `distance` of `early_stopping` is a function of the form `distance(best_score, score)`. By default `distance` is `-`, which implies that the monitored metric `f` is expected to be decreasing and mimimized. If you use some increasing metric (e.g. accuracy), you can customize the `distance` function: `(best_score, score) -> score - best_score`.

```julia
acc = let v = 0
  () -> v = max(1, v + 0.01)
end

> es = Flux.early_stopping(acc, 3; delta = (best_score, score) -> score - best_score)

Reviewer: ```suggestion
# early stopping trigger for acc with patience of 3
es = Flux.early_stopping(acc, 3; delta = (best_score, score) -> score - best_score)
```

# This will iterate until the 10th epoch
Flux.@epochs 10 begin
  es() && break
end

es = Flux.early_stopping(acc, 3)

# This will stop at the 3rd epoch
Flux.@epochs 10 begin
  es() && break

> end

Reviewer: Let's delete this second example in favor of my other suggestion with the loss example.

```

> 

Reviewer: ```suggestion

`early_stopping` and `plateau` are both built on top of `patience`. You can use `patience` to build your own triggers that use a patient counter. For example, if you want to trigger when the loss is below a threshold for several consecutive iterations:

```julia
threshold(f, thresh, delay) = Flux.patience(delay) do
  f() < thresh
end
```

```

Flux.stop
Flux.skip
```

## Patience Helpers

Flux provides utilities for controlling your training procedure according to some monitored condition and a maximum `patience`. For example, you can use `early_stopping` to stop training when the model is converging or deteriorating, or you can use `plateau` to check if the model is stagnating.

For example, below we create a pseudo-loss function that decreases, bottoms out, then increases. The early stopping trigger will break the loop before the loss increases too much.
```julia
# create a pseudo-loss that decreases for 4 calls, then starts increasing
# we call this like loss()
loss = let t = 0
  () -> begin
    t += 1
    (t - 4) ^ 2
  end
end

# create an early stopping trigger
# returns true when the loss increases for two consecutive steps
es = Flux.early_stopping(loss, 2; init_score=9)

# this will stop at the 6th (4 decreasing + 2 increasing calls) epoch

> Flux.@epochs 10 begin

Reviewer 2: I think it makes sense to drop the `Flux.`. I'm sure we can do a setup to import Flux and the symbols separately. I'd like to focus the API docs on the functions themselves.

Author: Do I also need to remove the `Flux.` in docstrings?

Reviewer: No those are necessary for the auto-docstrings to resolve.
